tests/regression/test/Prank.t.sol

```diff
@@ -48,43 +64,117 @@ contract PrankSetUpTest is Test {
 
 contract PrankTest is Test {
     Target target;
+    Target inner;
     Ext ext;
     Dummy dummy;
 
     function setUp() public {
         target = new Target();
+        inner = new Target();
+        target.setInnerTarget(inner);
         ext = new Ext();
     }
 
     function prank(address user) public {
         vm.prank(user);
     }
 
-    function check_prank(address user) public {
+    function checkNotPranked(Target _target, address realCaller) internal {
+        assertEq(_target.caller(), realCaller);
+        assertEq(_target.origin(), tx.origin);
+    }
+
+    function check_prank_single(address user) public {
         vm.prank(user);
+
+        // the outer call is pranked
         target.recordCaller();
         assert(target.caller() == user);
+        assert(target.origin() == tx.origin); // not pranked
 
+        // but the inner call is not pranked
+        checkNotPranked(inner, address(target));
+
+        // check that the prank is no longer active
         target.recordCaller();
-        assert(target.caller() == address(this));
+        checkNotPranked(target, address(this));
     }
 
-    function check_startPrank(address user) public {
+    function check_prank_double(address user, address origin) public {
+        vm.prank(user, origin);
+
+        // the outer call is pranked
+        target.recordCaller();
+        assert(target.caller() == user);
+        assert(target.origin() == origin);
+
+        // but the inner call is not pranked
+        checkNotPranked(inner, address(target));
```

Comment:
this also fails in foundry 🤨 is there an issue or is my mental model just wrong?

Reply:
update: my mental model was wrong, and so the implementation was wrong too. I checked that all the tests in PrankTest now behave the same in both foundry and halmos ([df2c7a3](https://github.com/a16z/halmos/pull/336/commits/df2c7a3e77afa3712f10d1d5e6d89e5f802b8fdf))